packages/monaco/src/browser/monaco-outline-contribution.ts

```diff
@@ -0,0 +1,108 @@
+* You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+*/
+
+import { injectable, inject } from "inversify";
+import SymbolInformation = monaco.modes.SymbolInformation;
+import SymbolKind = monaco.modes.SymbolKind;
+import { FrontendApplicationContribution, FrontendApplication } from "@theia/core/lib/browser";
+import { OutlineViewManager, OutlineSymbolInformation } from "@theia/outline-view/lib/browser/outline-view-manager";
+import { MonacoTextModelService } from './monaco-text-model-service';
+import { Range, EditorManager, EditorWidget } from '@theia/editor/lib/browser';
+import DocumentSymbolProviderRegistry = monaco.modes.DocumentSymbolProviderRegistry;
+import CancellationTokenSource = monaco.cancellation.CancellationTokenSource;
+import URI from "@theia/core/lib/common/uri";
+
+@injectable()
+export class MonacoOutlineContribution implements FrontendApplicationContribution {
+
+    ids: string[] = [];
+
+    constructor(
+        @inject(OutlineViewManager) protected readonly outlineViewManager: OutlineViewManager,
+        @inject(EditorManager) protected readonly editorManager: EditorManager,
+        @inject(MonacoTextModelService) protected readonly textModelService: MonacoTextModelService) { }
+
+    onStart(app: FrontendApplication): void {
+
+        this.outlineViewManager.onDidChangeOpenState(async isOpen => {
+            if (isOpen && this.editorManager.currentEditor) {
+                const model = await this.getModel(this.editorManager.currentEditor);
+                this.publish(model);
+            }
+        });
+
+        this.editorManager.onCurrentEditorChanged(async (editor: EditorWidget) => {
+            if (this.outlineViewManager.isOpen) {
+                const model = await this.getModel(editor);
+                this.publish(model);
+            }
+        });
+
+    }
+
+    protected async getModel(editor: EditorWidget): Promise<monaco.editor.IModel> {
+        const reference = await this.textModelService.createModelReference(editor.editor.uri);
+        const model = reference.object.textEditorModel;
+        model.onDidChangeContent(ev => {
+            this.publish(model);
+        });
+        return model;
+    }
+
+    protected async publish(model: monaco.editor.IModel) {
+        const entries: SymbolInformation[] = [];
+
+        const documentSymbolProviders = await DocumentSymbolProviderRegistry.all(model);
+
+        const cancellationSource = new CancellationTokenSource();
+        cancellationSource.cancel();
+        for (const documentSymbolProvider of documentSymbolProviders) {
+            const symbolInformation = await documentSymbolProvider.provideDocumentSymbols(model, cancellationSource.token);
+            entries.push(...symbolInformation);
+        }
+        const outlineSymbolInformations: OutlineSymbolInformation[] = this.getOutlineSymbolInformations(entries);
+        this.outlineViewManager.publish(outlineSymbolInformations);
```

Comment:
It should publish a tree not a list of symbols. I.e. the logic that computes the children based on ranges belongs here not in the outline view manager.